### PR TITLE
reworking of sorting function. Needs comments

### PR DESCRIPTION
**Pull request by clementstross**
Draft pull request! Do not approve!

### CODE REVIEW

**Review by CRSlade1 — APPROVED** [review]
Happy with this except few comments I've made. 
I've approved it for now, just so we can deal with the pull requests. 
You may need to update the lists/factor issue later.

File: R/prep_char_num_sort.R [hunk 1]
```diff
@@ -51,128 +51,136 @@ prep_char_num_sort <- function(array, NA_val="_NA", char_decreasing=FALSE, num_d
...
+    if (unique_val==TRUE){
+      return(unique(array_out))
+    }else{
+      return(array_out)
+    }
+  }
 
 
+  # Replace NAs and convert to string
+  array_na <- array %>% tidyr::replace_na(NA_val)
```
[the remark this answers is not in the document]
Comment (clementstross) [reply R/prep_char_num_sort.R:84]: Yeah I was very pleased when I decided to turn everything to a string and then undo it later. Saved me a lot of work

File: R/prep_char_num_sort.R [hunk 2]
```diff
@@ -51,128 +51,136 @@ prep_char_num_sort <- function(array, NA_val="_NA", char_decreasing=FALSE, num_d
...
-    return(if (unique_val) {array[1]} else {array})
-  } else {
-    # Find first character or numeric chunk. i.e. A_1_A would get A_.
-    first_element <- stringr::str_extract(array,"(\\-?[:digit:]+(\\.[:digit:]+)?)|([A-z\\.\\'@£$%~#&\\!\\?\\- ]+)")
-    # Conditions on NA_val:
-    # If character, needs to be added and sorted with characters.
-    # If numeric, needs to be added and sorted with numbers.
-    # If NA, needs to be added separately as NA.
+  # If Ordinal factor the sorting is simple
+  if (is.ordered(array)){
```
[the remark this answers is not in the document]
Comment (clementstross) [reply R/prep_char_num_sort.R:65]: I think forcing to an array is fine as a factor can be an array. I agree we don't want lists or dataframes

[the remark this answers is not in the document]
Comment (clementstross) [reply R/prep_char_num_sort.R:65]: I have just checked and it is ok. The check is added to the next commit

File: R/prep_char_num_sort.R [hunk 3]
```diff
@@ -51,128 +51,136 @@ prep_char_num_sort <- function(array, NA_val="_NA", char_decreasing=FALSE, num_d
...
-    # Handling characters
-    # Take location of characters and NAs if NA_val is character
-    char_index <- which(is.na(suppressWarnings(as.numeric(first_element))) & !is.na(first_element))
-    char_na_index <- if (is.character(NA_val) & !is.na(NA_val) & sum(is.na(array)) > 0) {which(is.na(array))} else {NULL}
-    full_char_index <- c(char_na_index, char_index)
+  array_out <- c()
+  # Sort parts starting with a string
+  # Find string until a number if there is one
+  if (length(start_char) > 0){
+    start_char_char_part <- start_char %>% gsub("(^\\D+)(-?\\d+\\.?\\d*)(\\D*)(.*)", "\\1", .) %>% unique()
```
[the remark this answers is not in the document]
Comment (clementstross) [reply R/prep_char_num_sort.R:103]: Done

File: R/prep_char_num_sort.R [hunk 4]
```diff
@@ -51,128 +51,136 @@ prep_char_num_sort <- function(array, NA_val="_NA", char_decreasing=FALSE, num_d
...
+  checkmate::assert_true(setequal(array_out,array_na))
+
+  # if (any(is.na(suppressWarnings(as.numeric(array_out))))){
+  #   return(array_out)
+  # }else{
+  #   return(array_out %>% as.numeric())
+  # }
+
+  return(array_out)
 }
```
[the remark this answers is not in the document]
Comment (clementstross) [reply R/prep_char_num_sort.R:186]: I think we should raise and issue/feature enhancement.

**Pull request merged**